### PR TITLE
Unfilter `xtokens.transfer_multicurrencies` and bump `MaxAssetsForTransfer` to 2

### LINKED ISSUE #516: Integrate Statemine assets
- [x] Test locally
- [x] Test on Rococo
- [x] Send message to request to open Calamari -> Statemine channel
- [x] Governance motion on Kusama to command Statemine to accept our channel and open Statemine -> Calamari
- [ ] Send message to accept their channel
- [ ] Register RMRK and USDT assets
- [x] Enable `xtokens.transferMulticurrencies`

USDT: 
`parents: 1 interior: { X3: [ { Parachain: 1,000 } { PalletInstance: 50 } { GeneralIndex: 1,984 } ] }`
```
  deposit: 0
  name: Tether USD
  symbol: xcUSDT
  decimals: 6
  isFrozen: false
  minBalance: 1
  isSufficient: true
```
RMRK:
`parents: 1 interior: { X3: [ { Parachain: 1,000 } { PalletInstance: 50 } { GeneralIndex: 8 } ] }`
```
  deposit: 0
  name: xcRMRK
  symbol: xcRMRK
  decimals: 10
  isFrozen: false
  minBalance: 1
  isSufficient: true
```

Example governance motion ----> https://kusama.polkassembly.io/motion/417

### PR DESCRIPTION
**Pull request by ghzlatarev**
Signed-off-by: Georgi Zlatarev <georgi.zlatarev@manta.network>

## Description

closes: #516 

* Statemine currently only allows to pay fees in KSM, so if we want to send assets to Statemine we need to also send KSM.
* To do so we have to unfilter `xtokens.transfer_multicurrencies` 
* Also have to bump `MaxAssetsForTransfer` to 2
* Added some must-fail and must pass tests to simulate different scenarios.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [x] Verify benchmarks & weights have been updated for any modified runtime logics
- [x] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [x] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [x] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: runtime/common/Cargo.toml
Comment (ghzlatarev) [inline runtime/common/Cargo.toml:16]: TODO: remove these.

File: runtime/common/tests/xcm_mock/parachain.rs
Comment (ghzlatarev) [inline runtime/common/tests/xcm_mock/parachain.rs]: Need 3 for some of the must-fail scenarios.

File: runtime/calamari/src/lib.rs
Comment (Garandor) [inline runtime/calamari/src/lib.rs]: this'll need redoing after we merge #590 as the lines moved

**Comment by Garandor** [comment]
merge conflicts to resolve

File: runtime/calamari/src/lib.rs
Comment (Garandor) [inline runtime/calamari/src/lib.rs]: minor: i'm surprised this passes `cargo fmt`, could you please add the second one on a new line like CollatorSelection above

File: runtime/dolphin/src/lib.rs
Comment (Garandor) [inline runtime/dolphin/src/lib.rs]: if you prefer the condensed look without a newline, pls also adapt manta and calamari so we dont diverge

File: runtime/dolphin/src/lib.rs
Comment (Garandor) [inline runtime/dolphin/src/lib.rs]: same

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [inline runtime/common/tests/xcm_tests.rs:2209]: how come the locations are identical between paraA and B? I thought XCM locations are relative to the source?

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [inline runtime/common/tests/xcm_tests.rs:2257]: b was registered but never minted a balance. is the balance transferred here from the genesis endowment to alice on mock net?

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [inline runtime/common/tests/xcm_tests.rs:2316]: according to your comment above this should be `relay_asset_amount_minted_on_a - min_xcm_fee`

Which one is correct?

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [inline runtime/common/tests/xcm_tests.rs:2459]: I'm not sure what you're testing here. Could you add a `///` doccomment fn header describing the test scenarios with a few words?

**Review by Garandor — COMMENTED** [review]
minor format notes and questions.

Also, thanks for using clear variable names

Gonna leave approval to @Dengjianping and @stechu, not comfortable approving xcm code yet.

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2209]: Yes you're right, the correct way of representing a self location (native asset location) should be (0, Here). 
However representing in a "absolute" way (1, X1(self_para_id)) was a common mistake that many teams copied, including us. 
We worked around this mistake `impl<T, Balance> MatchesFungible<Balance> for IsNativeConcrete<T>` where we also check for `MultiLocation::is_here` 
Actually my next PR will be to actually start using the correct way of handling self locations and the code is ready, just need the other PRs to go in first. https://github.com/Manta-Network/Manta/issues/452

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2257]: Yeah this is in the mock setup:
```
pallet_balances::GenesisConfig::<Runtime> {
		balances: vec![(ALICE, INITIAL_BALANCE)],
	}
	.assimilate_storage(&mut t)
	.unwrap();
```
This is kind of hidden logic , which I'm usually against , so I didn't for example add `AssetManager` storage setup there. But this is not too unreasonable to assume, that's why I didn't rewrite it.
Not a big deal if you guys want to make it explicit, just let me know.

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2316]: The comment is not great. The total supply should be counted from both chains:
`relay_asset_amount_minted_on_a - fee_amount` is still on `ParaA`
`fee_amount - min_xcm_fee` is on ParaB.
And the thing to notice is that it comes out to `relay_asset_amount_minted_on_a - min_xcm_fee` meaning 1 `min_xcm_fee` is destroyed. This is a design choice by ORML people to make these kind of transfer. The end situation is that some of the relay chain token held as reserve on the relay chain, will become unwithdrawable.
Will rewrite it.

File: runtime/calamari/src/lib.rs
Comment (ghzlatarev) [reply runtime/calamari/src/lib.rs]: Oops forgot about this. I noticed that this wasn't getting formatted with ctrl-s, weird...

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2459]: First part are supposed to be only must fail cases on the sender side. covering all the error messages.
Added some comments on the asserts.
The second part are must fail case on the receiver part.

File: runtime/dolphin/src/lib.rs
Comment (ghzlatarev) [reply runtime/dolphin/src/lib.rs]: No, accidentally removed this line. Reverted it now.

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2459]: Please don't make me write comments for all the tests though :D

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [reply runtime/common/tests/xcm_tests.rs:2316]: So then the comment is correct and the logic in 2311 isn't, right?

P.S. since you checked in changes for the other points but not this one, am I right to assume you forgot to add this change?
You might want to resolve a comment only after you've pushed the commit that makes it outdated. I use that myself as kind of a checklist

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [reply runtime/common/tests/xcm_tests.rs:2459]: Agree it'd be torture to comment each test case, my idea was to annotate the main fn:
https://github.com/Manta-Network/Manta/pull/576/files/17def4e433d575d75d82e3de7150e0871ce4ae3d#diff-c55c7706ced753270f03a35c10fdb301207231f2224a8e3f3951c5970f546643R2365

With a comment describing the structure you just put in the comment here.
So that when i'm reading the code I know what to expect from this long fn instead of having to rhyme things together line by line.

An alternative would be to split this up into multiple shorter fns (which I'd actually prefer, but might not be reasonable)

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2316]: Hmm something weird happened here, I wrote a big comment but it wasn't pushed.

Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2316]: https://github.com/Manta-Network/Manta/pull/576/commits/a140e5a016635b48e2abcf985d84d2e98461b4fc This is the commit with the change, I don't know why it's not reflected in this diff.
Also I remember resolving this conversation, but somehow it is not resolved...

Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2316]: Actually none of my changes from the last 3 commits are reflected

Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2316]: Please refresh the diff, should be showing up correctly now , but I think I understood your question about 2311:
```
                               (
					parachain::CurrencyId::MantaCurrency(relay_asset_id_on_a),
					fee_amount
				)
```
This is the fee being sent to ParaB, but the logic in ORML will subtract `min_xcm_fee` so the comment is `fee_amount - min_xcm_fee is on ParaB.` is correct. Let me write this explicitly.

File: runtime/common/tests/xcm_tests.rs
Comment (ghzlatarev) [reply runtime/common/tests/xcm_tests.rs:2459]: Ok, added the comment. Let me know if it doesn't make sense.

File: runtime/common/tests/xcm_tests.rs
Comment (Garandor) [reply runtime/common/tests/xcm_tests.rs:2316]: ok, got it thanks. i thought total_supply was calculated over both chains, not from the view of ParaA

Also, I unresolved this comment earlier to make sure you see it before merging

**Review by Garandor — DISMISSED** [review]
LGTM after merging the log conflict

**Review by Garandor — DISMISSED** [review]

**Review by Dengjianping — DISMISSED** [review]
LTGM!
Thanks for the PR.
But I think this extrinsic `transfer_multicurrencies` is kind of more complicated than `transfer`, it would be better if there's a how-to tutorial.

**Comment by ghzlatarev** [comment]
> LTGM! Thanks for the PR. But I think this extrinsic `transfer_multicurrencies` is kind of more complicated than `transfer`, it would be better if there's a how-to tutorial.

I think we decided not to advertise the polkadot-js front-end to users at all, otherwise I would've written docs for the `transfer` as well. We want to push users to our app, that was the reason.

**Review by Dengjianping — DISMISSED** [review]
LGTM!

**Review by Garandor — DISMISSED** [review]

**Review by Dengjianping — APPROVED** [review]
LGTM!

**Review by Garandor — APPROVED** [review]

**Pull request merged**
